nixos/nextcloud: Expose more config.php options via Nix module

aidalgol opened the pull request:
###### Description of changes

Expose more Nextcloud `config.php` options through the Nix module:
- `default_locale`
- 2FA settings
- mail settings

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

aidalgol committed:
nixos/nextcloud: Expose the "default_locale" option

Add an option to the nix module to set the "default_locale" config.php option.

aidalgol committed:
nixos/nextcloud: Expose 2FA options

Allow setting the two-factor authentication options via nix options instead of
letting Nextcloud manage these.

aidalgol committed:
nixos/nextcloud: Expose mail options

Allow setting the email options via nix options instead of letting Nextcloud
manage these.

danth commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -374,6 +381,222 @@ in {
+
+            Whether to enforce two-factor authentication.
+
+            Further details about this setting and supported locale codes can be found in the
+            <link xlink:href="${twoFactorAuthUpstreamManualUrl}">upstream documentation</link>.
```
Perhaps define this entire paragraph as a variable rather than just the URL, since it is all repeated.

danth commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -374,6 +381,222 @@ in {
+
+            The mail server mode.
+
+            Further details about this setting can be found in the
+            <link xlink:href="${mailUpstreamManualUrl}">upstream documentation</link>.
```
Perhaps define this entire paragraph as a variable rather than just the URL, since it is all repeated.

To make the code even dryer, you could define a function which takes the main paragraph as an argument:

```nix
mainDescription: ''
  ${uiClashWarning}

  ${mainDescription}

  Further details about this setting can be found in the
  <link xlink:href="https://docs.nextcloud.com/server/stable/admin_manual/configuration_server/email_configuration.html">upstream documentation</link>.
''
```

aidalgol replied:
I actually wanted to refactor this "Further details about this setting" paragraph from every option description that has it, but I thought it would be out of scope for this PR.  Are you OK with leaving this as is in this changeset if I followup with another PR?

danth replied:
Yeah. It's probably better to do the refactoring separately, that way we can get this merged sooner.

danth commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -268,7 +268,14 @@ in {
+      uiClashWarning = ''
+        <warning>
+          <para>Setting this option will prevent changing this via the
+          Nextcloud UI, but it wil still appear editable in the UI.</para>
```
```suggestion
          Nextcloud UI, but it will still appear editable in the UI.</para>
```

aidalgol replied:
Fixed.

danth commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -374,6 +381,222 @@ in {
+        example = "de_CH";
+        description = ''
+          <warning>
+           <para>This option exists since Nextcloud 21! If older versions are used,
+            this will throw an eval-error!</para>
```
```suggestion
           <para>This option has only existed since Nextcloud 21! If an older version
            of Nextcloud is used, this will throw an error.</para>
```

aidalgol replied:
I have opted to instead remove this paragraph, as Nextcloud 22 is now the earliest version available in nixpkgs.

aidalgol committed:
nixos/nextcloud: Rename option from code-review feedback

Rename "twofactorEnforced.enabled" to "twofactorEnforcement.enable"

aidalgol committed:
nixos/nextcloud: Fix typo

"Will" was spelt "wil".

aidalgol committed:
nixos/nextcloud: Remove obsolete warnings

Remove the warning in defaultPhoneRegion and defaultLocale about the options
only existing since Nextcloud 21, since Nextcloud 22 is now the earliest version
available in nixpkgs.

aidalgol commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -374,6 +381,222 @@ in {
+      twofactorEnforced = let
+        twoFactorAuthUpstreamManualUrl = "https://docs.nextcloud.com/server/latest/admin_manual/configuration_user/two_factor-auth.html";
+      in {
+        enabled = mkOption {
```
Renamed.

Ma27 commented:
SGTM, but before merging I'd need some time for some local tests.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -374,6 +376,217 @@ in {
+      twofactorEnforcement = let
+        twoFactorAuthUpstreamManualUrl = "https://docs.nextcloud.com/server/latest/admin_manual/configuration_user/two_factor-auth.html";
+      in {
+        enable = mkOption {
```
This one is somewhat dangerous: when enabling this straight from the beginning, `root` cannot authenticate because of the 2fa enforcement. This should be reflected here.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -374,6 +376,217 @@ in {
+      in {
+        mode = mkOption {
+          default = null;
+          type = types.nullOr (types.enum [ "smtp" "sendmail" "qmail" ]);
```
hmm even when specifying `sendmail` here in a Nix expression for nextcloud, the admin UI shows me `smtp`(`config.php` and `override.config.php` are correct though). Any idea what's up here?

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -268,7 +268,14 @@ in {
+      uiClashWarning = ''
+        <warning>
+          <para>Setting this option will prevent changing this via the
+          Nextcloud UI, but it will still appear editable in the UI.</para>
```
Actually it doesn't: when messing around in the mail settings, stuff like `mail_domain`/`mail_from_address`/`mail_sendmailmode` appear with the values from the UI in `config.php`. Yes, things from `override.config.php` take precedence, but since all of these settings are optional, this can still cause unwanted side-effects if parts of the config come from the UI and another part from the Nix expression.

Suggestions what to do:

* Force all `mail`-options to be declared if enabled
* Perhaps even write a patch to disable the controls in the UI.

To be frank, I'm not sure anymore if that's a reasonable thing to do. cc @aanderse @nlewo @ajs124 @beardhatcode for opinions.

aanderse replied:
Some software really wants to be configured via the ui. We do something similarly hacky with the [nzbget](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/nixos/modules/services/misc/nzbget.nix) module... but the module is tiny, the code is easy to understand, and overall it is a pretty simple piece of software compared to `nextcloud`. Not directly comparable, but maybe my comments provide at least a little bit of value.

:man_shrugging:

Ma27 commented:
@aanderse is probably right, this would blow up the code even more and as I said, to correctly implement that you'd have to patch nextcloud.
If people want to do that anyways, it should be done via #118093. However I think that using the UI (or perhaps terraform with a suitable provider) is the better choice for this kind of state management.
